source/Mlos.Python/mlos/Optimizers/unit_tests/TestParetoFrontier.py

```diff
@@ -0,0 +1,336 @@
+        #
+        #   assert y0**2 + y1**2 + ... == radius**2
+        #
+        assert (np.power(objectives_df, 2).sum(axis=1) - np.power(random_params_df["spherical_coordinates.radius"], 2) < 0.000001).all()
+
+
+        # Just a few more sanity checks before we do the pareto computation.
+        #
+        if minimize == "all":
+            assert (objectives_df <= 0).all().all()
+        elif minimize == "none":
+            assert (objectives_df >= 0).all().all()
+        else:
+            for column, minimize_column in zip(objectives_df, minimize_mask):
+                if minimize_column:
+                    assert (objectives_df[column] <= 0).all()
+                else:
+                    assert (objectives_df[column] >= 0).all()
+
+
+        pareto_df = ParetoFrontier.compute_pareto(
+            optimization_problem=optimization_problem,
+            objectives_df=objectives_df
+        )
+
+        # We know that all of the pareto efficient points must be on the frontier.
+        #
+        assert optimal_points_index.difference(pareto_df.index.intersection(optimal_points_index)).empty
+        assert len(pareto_df.index) >= len(optimal_points_index)
+
+        # If we flip all minimized objectives, we can assert on even more things.
+        #
+        for column, minimize_column in zip(objectives_df, minimize_mask):
+            if minimize_column:
+                objectives_df[column] = -objectives_df[column]
+                pareto_df[column] = - pareto_df[column]
+
+        non_pareto_index = objectives_df.index.difference(pareto_df.index)
+        for i, row in pareto_df.iterrows():
+            # Now let's make sure that no point in pareto is dominated by any non-pareto point.
+            #
+            assert (objectives_df.loc[non_pareto_index] < row).any(axis=1).sum() == len(non_pareto_index)
+
+            # Let's also make sure that no point on the pareto is dominated by any other point there.
+            #
+            other_rows = pareto_df.index.difference([i])
+            assert (pareto_df.loc[other_rows] > row).all(axis=1).sum() == 0
+
+
+    def test_repeated_values(self):
+        """Validates that the algorithm does its job in the presence of repeated values.
+
+        :return:
+        """
+
+        optimization_problem = OptimizationProblem(
+            parameter_space=None,
+            objective_space=SimpleHypergrid(
+                name="objectives",
+                dimensions=[
+                    ContinuousDimension(name='y1', min=0, max=5),
+                    ContinuousDimension(name='y2', min=0, max=5)
+                ]
+            ),
+            objectives=[
+                Objective(name='y1', minimize=False),
+                Objective(name='y2', minimize=False)
+            ]
+        )
+
+        expected_pareto_df = pd.DataFrame(
+            [
+                [1, 2],
+                [1, 2],
+                [2, 1],
+                [0.5, 2],
+                [1, 1],

```

Comment:
I'm not sure I understand this. Shouldn't [1, 2] and [2, 1] dominate all the other points?